### ISSUE TITLE
[ROCm] `nn.Embedding(sparse=True)` backward computes ~10^5x incorrect gradient magnitude on ROCm (gfx1201 / RDNA4)

### ISSUE THREAD

**Issue by josharmour** [pytorch/pytorch#193704]
### 🐛 Bug

When using `nn.Embedding(..., sparse=True)`, running `.backward()` on ROCm (tested on AMD Radeon AI PRO R9700 / `gfx1201` RDNA4 with PyTorch ROCm) produces gradients whose values are grossly inaccurate—approximately **~100,000x** larger in magnitude than expected CPU / reference CUDA gradients.

Dense embeddings (`sparse=False`) produce correct gradients on the exact same ROCm device, isolating the bug specifically to the HIP sparse embedding backward kernel / accumulation path (`embedding_backward` sparse kernel).

### Versions

- PyTorch: 2.x+rocm6.x
- OS: Linux x86_64
- GPU: AMD Radeon AI PRO R9700 (`gfx1201` / RDNA4)

### Minimal Reproducible Example

```python
import torch
import torch.nn as nn

torch.manual_seed(42)
indices = torch.randint(0, 100_000, (4, 16))

# 1. CPU Reference
emb_cpu = nn.Embedding(100_000, 64, sparse=True)
loss_cpu = emb_cpu(indices).sum()
loss_cpu.backward()
grad_cpu = emb_cpu.weight.grad.coalesce()

# 2. ROCm Device
device = "cuda"  # ROCm HIP device
emb_gpu = nn.Embedding(100_000, 64, sparse=True).to(device)
with torch.no_grad():
    emb_gpu.weight.copy_(emb_cpu.weight)

loss_gpu = emb_gpu(indices.to(device)).sum()
loss_gpu.backward()
grad_gpu = emb_gpu.weight.grad.coalesce()

# Compare
mean_cpu = grad_cpu.values().abs().mean().item()
mean_gpu = grad_gpu.values().cpu().abs().mean().item()
print(f"CPU mean grad: {mean_cpu:.6f}")
print(f"GPU mean grad: {mean_gpu:.6f}")
print(f"Ratio (GPU / CPU): {mean_gpu / mean_cpu:.2f}x")
```

### Expected Behavior

Output should show `Ratio (GPU / CPU): 1.00x`.

### Actual Behavior

```text
CPU mean grad: 1.000000
GPU mean grad: 99998.453125
Ratio (GPU / CPU): 99998.45x
```

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip @ezyang @albanD @gqchen @soulitzer @Varal7 @bobrenjc93 @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

**Comment by doplxyz** [comment]
### Not reproduced on ROCm 7.2.4 / 7.14 with gfx1201 (10 runs)

I have an RX 9070 XT (gfx1201, RDNA4) and ran the script from the opening post verbatim
(unmodified, sha256 `a8fd8c28236f7e37d9a00e0a5da2997514a7301bccd2c8a2a90c6e572befd75e`),
in 5 separate processes on each of two published images:

| | torch | HIP | image digest |
|---|---|---|---|
| A | `2.12.0+rocm7.14.0` | `7.14.60850` | `rocm/pytorch@sha256:c38eeda81d85f00fbe35d3d50ce42ce59c524e87d810624f4eb5c52fddb3b9ad` |
| B | `2.9.1+rocm7.2.4.git39497456` | `7.2.53211-97f5574fe2` | `rocm/pytorch@sha256:7fe531fa185af260352fe7fbb3fa64ad749abe72adf0600a648c4692801b125a` |

All 10 runs printed `GPU mean grad: 1.000000` / `Ratio (GPU / CPU): 1.00x`.
`PYTORCH_ALLOC_CONF` was unset in both images, and no other process held the GPU.

I have no information about the `2.x+rocm6.x` build in the report — I do not have it and the
commit is not stated — so this is **not** a claim that the original observation was wrong.
It is only: not reproduced in 10 runs on these two stacks.

#### On the attribution in the opening post

> isolating the bug specifically to the HIP sparse embedding backward kernel

In the two images above, the only dispatch header shipped for this operator is
`ATen/ops/embedding_sparse_backward_compositeimplicitautograd_dispatch.h`, while
`embedding_dense_backward` ships both `_cpu_dispatch.h` and `_cuda_dispatch.h`. I did not read
the generated registration TUs, so I will put it no more strongly than that: **in the shipped
dispatch headers I checked, there is no backend-specific entry for the sparse path.**

Consistent with that, `aten/src/ATen/native/Embedding.cpp` is byte-identical between the two
commits these images were built from (`13eab2b3ef4e` and `394974567f84`). For the argument values
this repro uses (`padding_idx=-1`, `scale_grad_by_freq=False`), `embedding_sparse_backward`
reshapes the indices and calls `_sparse_coo_tensor_unsafe_symint`, i.e. it returns an
**uncoalesced** tensor. (With other arguments it also filters `padding_idx` rows and rejects
`scale_grad_by_freq`.)

I also checked directly: after `backward()`, the raw uncoalesced values are all exactly `1.0`
here, at `D`=64, 256 and 512. The backward output itself is fine. The backend-specific work that
this repro reaches is the `.coalesce()` it calls afterwards.

#### A separate defect in `coalesce()`, which does **not** explain this issue

While checking that step I found that `coalesce_sparse_cuda` can leave part of its output
unwritten on gfx1201, from a grid-index mismatch introduced in #158281. I have posted the root
cause, the measurements and a one-line fix on **#158281**, not here, because it is a different
defect.

**It cannot be the cause of the symptom reported here.** It requires `gridDim.z > 1`, i.e.
`D > warp_size*4` (128 on wave32); this repro uses `D=64`, so `blockIdx.z` is always 0 and the
faulty term is multiplied by zero. My `D=64` runs above are clean for that reason, and I am not
offering it as an explanation for the `rocm6.x` observation.

One methodological note from that work does carry over, though, if anyone tries to reproduce
**this** issue: the corrupted region there is uninitialized memory, and the caching allocator
frequently hands back a block that already contains the expected value. Repeating one failing
`coalesce()` case 10 times gave me alternating results (`[64, 0, 64, 0, ...]` bad columns) purely
from allocator reuse. So for any uninitialized-memory-shaped symptom, a single clean run is weak
evidence. If the `rocm6.x` environment still exists, running it several times may be worthwhile.

Happy to run anything else on gfx1201 if it helps.

**Comment by josharmour** [comment]
Closing as verified not reproducible.

Retested across multiple runs on the original machine / hardware (`gfx1201` / AMD Radeon AI PRO R9700 on PyTorch 2.9.1+rocm6.4 / HIP 6.4) and confirmed that the verbatim repro from the issue description outputs expected gradients (`Ratio (GPU / CPU): 1.00x`).

Thanks @doplxyz for the detailed analysis and clarification regarding `coalesce_sparse_cuda` in #158281.

**Comment by josharmour** [comment]
Thanks @doplxyz for the thorough testing and analysis on gfx1201.

To clarify the exact environment where this originally occurred:
- **PyTorch**: `2.9.1+rocm6.4`
- **HIP**: `6.4.43484-123eb5128`
- **GPU**: AMD Radeon AI PRO R9700 (`gfx1201` / RDNA4)

The non-deterministic behavior and allocator buffer reuse you described matches what we observed under ROCm 6.4 (which produced inflated gradient magnitudes / NaNs during sparse backward accumulation).

It looks like the newer ROCm 7.2 / 7.14 stacks have resolved this on gfx1201. For our pipeline on ROCm 6.4, routing the sparse embedding autograd through CPU (`hybrid_cpu_embedding = True`) completely bypassed the kernel issue.

Happy to leave this closed since ROCm 7.x handles this cleanly.
